src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+package com.fulcrumgenomics.umi
+
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicLong
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sequences, SimpleCounter}
+import com.fulcrumgenomics.util.Sequences.countMismatches
+import dagr.commons.CommonsDef.{DirPath, PathToBam}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.cmdline.ValidationException
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools._
+import htsjdk.samtools.util.SortingCollection
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+
+object GroupReadsByUmi {
+  private val SortPositionAttribute        = "__GRs"
+  private val LibraryAttribute             = "__GRlb"
+  private val InsertEndsAndStrandAttribute = "__Insert"
+
+  /** Trait that can be implemented to provide a UMI assignment strategy. */
+  sealed trait UmiAssigner {
+    def assign(rawUmis: Seq[String]) : Map[String, String]
```

Comment:
thinking out loud, is there every a time we would want to not return an assignment for a particular `rawUmi`?   If so, then we should state that there is no guarantee that the resulting map contains all the `rawUmi`s.